src/video/encoder.cpp

```diff
@@ -0,0 +1,120 @@
+#include "encoder.hpp"
+
+#include <iostream>
+
+extern "C" {
+#include <x264.h>
+
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavformat/avio.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+}
+
+namespace h264encoder {
+
+class EncoderImpl {
+private:
+	int in_xres, in_yres, out_xres, out_yres;
+	int framecounter;
+	int nheader;
+	x264_t* enc;
+	x264_param_t prms;
+	x264_picture_t pic_in, pic_out;
+	x264_nal_t* nals;
+
+	struct SwsContext* sws;
+	AVFrame pic_raw; /* used for our "raw" input container */
+	AVPixelFormat cam_pixel_fmt = AV_PIX_FMT_BGR24;
+
+public:
+	std::vector<x264_nal_t_simple> nals_simple;
+	int num_nals;
+	virtual ~EncoderImpl() {}
+	EncoderImpl(){};
+
+	// Setup of the encoder instance
+	EncoderImpl(int inW, int inH, int outW, int outH, float fps)
+		: in_xres(inW), in_yres(inH), out_xres(outW), out_yres(outH) {
+
+		framecounter = 0;
+		x264_param_default_preset(&prms, "ultrafast", "zerolatency,fastdecode");
+		x264_param_apply_profile(&prms, "baseline");
+		prms.i_width = out_xres;
+		prms.i_height = out_yres;
+		prms.i_fps_num = fps;
+		prms.i_fps_den = 1;
+		prms.rc.i_qp_constant = 51; // maybe try 0 if it is lagging, max 51
+
+		prms.rc.i_rc_method = X264_RC_CRF;
+		prms.rc.f_rf_constant = 25; // maybe try 51 if it is lagging?
+		prms.rc.f_rf_constant_max = 25;
```

Comment:
Can we make `qp_constant` and `rf_constant` parameters to the class? Not great to have hardcoded magic values that would affect the stream resolution